File: apps/api/app/schemas/papers.py
"""Paper & Annotation schemas."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class PaperCreate(BaseModel):
    project_id: str
    title: str
    authors: list[dict[str, Any]] | None = None
    abstract: str | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    pmid: str | None = None
    year: int | None = None
    pdf_path: str | None = None
    metadata_json: dict[str, Any] | None = None
    extraction_status: str = "ok"


class PaperResponse(BaseModel):
    id: str
    project_id: str
    title: str
    authors: list[dict[str, Any]] | None = None
    abstract: str | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    year: int | None = None
    extraction_status: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaperDetailResponse(BaseModel):
    id: str
    project_id: str
    title: str
    authors: list[dict[str, Any]] | None = None
    abstract: str | None = None
    doi: str | None = None
    arxiv_id: str | None = None
    pmid: str | None = None
    year: int | None = None
    metadata_json: dict[str, Any] | None = None
    extraction_status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PaperStatusResponse(BaseModel):
    paper_id: str
    step: (
        str  # Currently always 'ready' -- see audit-11 H-5; real transitions need a background queue
    )
    step_index: int  # Currently always 4
    extraction_status: str  # 'ok' | 'unverified'
    chunks_count: int = 0
    message: str | None = None


class AnnotationCreate(BaseModel):
    paper_id: str
    page_number: int = 1
    selected_text: str
    highlight_color: str = "yellow"
    note_text: str | None = None
    position_data: dict[str, Any] | None = None


class AnnotationUpdate(BaseModel):
    highlight_color: str | None = None
    note_text: str | None = None
    ai_thread: list[dict[str, Any]] | None = None


class AnnotationResponse(BaseModel):
    id: str
    paper_id: str
    user_id: str
    page_number: int
    selected_text: str
    highlight_color: str
    note_text: str | None = None
    ai_thread: list[dict[str, Any]] | None = None
    position_data: dict[str, Any] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: apps/web/src/lib/api/papers.ts
import { request, resolveApiUrl, ApiError, extractErrorMessage } from './client';
import type { Author, ExtractionStatus } from '@openresearch/citations';

export interface PaperDTO {
  id: string;
  project_id: string;
  title: string;
  authors?: Author[];
  abstract?: string;
  doi?: string;
  arxiv_id?: string;
  pmid?: string;
  year?: number;
  pdf_path?: string;
  extraction_status: ExtractionStatus;
  metadata_json?: Record<string, unknown>;
  created_at: string;
  updated_at?: string;
}

export interface PaperAnnotationDTO {
  id: string;
  paper_id: string;
  user_id: string;
  page_number: number;
  selected_text: string;
  highlight_color: string;
  note_text?: string;
  ai_thread?: Array<{ role: 'user' | 'assistant'; message: string; timestamp: string }>;
  position_data?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface PaperUploadResponseDTO extends PaperDTO {
  indexed_chunks?: number;
}

export interface PaperStatusDTO {
  id: string;
  title: string;
  extraction_status: ExtractionStatus;
  chunks_count?: number;
  annotations_count: number;
}

export const papersApi = {
  list: (projectId: string, query?: string) => {
    const q = query ? `?q=${encodeURIComponent(query)}` : '';
    return request<PaperDTO[]>(`/projects/${projectId}/papers${q}`);
  },
  get: (id: string) => request<PaperDTO>(`/papers/${id}`),
  upload: async (projectId: string, file: File): Promise<PaperUploadResponseDTO> => {
    const formData = new FormData();
    formData.append('file', file);

    const url = resolveApiUrl();
    const response = await fetch(`${url}/projects/${projectId}/papers/upload`, {
      method: 'POST',
      body: formData,
    });

    if (!response.ok) {
      throw new ApiError(
        await extractErrorMessage(response, 'Paper upload failed'),
        response.status
      );
    }

    return response.json();
  },
  status: (id: string) => request<PaperStatusDTO>(`/papers/${id}/status`),
  delete: (id: string) =>
    request<void>(`/papers/${id}`, {
      method: 'DELETE',
    }),
  getAnnotations: (paperId: string) => request<PaperAnnotationDTO[]>(`/papers/${paperId}/annotations`),
  createAnnotation: (
    paperId: string,
    data: {
      page_number: number;
      selected_text: string;
      highlight_color?: string;
      note_text?: string;
      position_data?: Record<string, unknown>;
    }
  ) =>
    request<PaperAnnotationDTO>(`/papers/${paperId}/annotations`, {
      method: 'POST',
      body: JSON.stringify({ paper_id: paperId, ...data }),
    }),
  updateAnnotation: (
    paperId: string,
    annotationId: string,
    data: {
      highlight_color?: string;
      note_text?: string;
      ai_thread?: Array<{ role: 'user' | 'assistant'; message: string; timestamp: string }>;
    }
  ) =>
    request<PaperAnnotationDTO>(`/papers/${paperId}/annotations/${annotationId}`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    }),
  deleteAnnotation: (paperId: string, annotationId: string) =>
    request<void>(`/papers/${paperId}/annotations/${annotationId}`, {
      method: 'DELETE',
    }),
  ask: (paperId: string, data: { selected_text?: string; page_number?: number; question?: string; prompt_type?: string }) =>
    request<{ answer: string; prompt_type: string; grounded?: boolean; sources?: string[] }>(`/papers/${paperId}/ask`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
};
